### PR TITLE
Allow docker-nginx to run under a non root user

### PR DESCRIPTION
**Pull request by sravankasu**

### CODE REVIEW

**Comment by sravankasu** [comment]
@bryanlatten Please review

File: Dockerfile [hunk 1]
```diff
@@ -54,6 +54,8 @@ RUN sed -i "s/listen [0-9]*;/listen ${CONTAINER_PORT};/" $CONF_NGINX_SITE && \
     sed -i "s/}/\n    font\/woff                             woff;&/" /etc/nginx/mime.types && \
     sed -i "s/}/\n    font\/woff2                            woff2;\n&/g" /etc/nginx/mime.types
 
+RUN /bin/bash -e /scripts/set_permissions.sh
```
Comment (bryanlatten) [inline Dockerfile]: does this need to be its own distinct layer? is it possible to combine with another?

File: container/root/s6-setuidgid [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+current_user=$(whoami)
```
Comment (bryanlatten) [inline container/root/s6-setuidgid]: this seems like something we might want to maintain in the base, no? 

no immediate action, but let's get this out, then uplevel to docker-base after

File: Dockerfile (hunk 1 above)
Comment (sravankasu) [reply Dockerfile]: No it does not, shall I combine it with the layer above this?

File: container/root/s6-setuidgid (hunk 2 above)
Comment (sravankasu) [reply container/root/s6-setuidgid]: Yes, I was actually debating that but was not sure if we were ready to make changes to the base yet.

File: Dockerfile (hunk 1 above)
Comment (bryanlatten) [reply Dockerfile]: Yes please!

Comment (sravankasu) [reply Dockerfile]: Done

**Comment by bryanlatten** [comment]
@sravankasu last change required - can you update the expectations for the README for adding any code to the container?

**Comment by sravankasu** [comment]
@bryanlatten How does this sound?

"Any new script/file that needs to be added must be given proper permissions/ownership to the non root user through `container/root/scripts/set_permissions.sh`. This is to ensure that the image can be run under a non root user."

**Comment by sravankasu** [comment]
Also I'm seeing the following message when running the image under a non root user:
`nginx: [alert] could not open error log file: open() "/var/lib/nginx/logs/error.log" failed (13: Permission denied)`

This file seems to be generated by the nginx service after it has started, so not sure how to handle this. 

However I see that we're printing error logs to stdout [here](https://github.com/behance/docker-nginx/blob/master/container/root/etc/nginx/nginx.conf#L17) (or atleast I think I do), so maybe this is a non issue?

**Comment by bryanlatten** [comment]
@sravankasu what non-root user are you running nginx as? The workers are pre-set to work as `www-data` (nginx standard), the master likely needs to match that user

**Comment by sravankasu** [comment]
I'm using the `NOT_ROOT_USER` env var set in the Dockerfile, so it's the same user.

**Comment by bryanlatten** [comment]
https://stackoverflow.com/questions/25774999/nginx-stat-failed-13-permission-denied

**Comment by sravankasu** [comment]
@bryanlatten I chose to create an empty file with the same name in the same path before which I could assign ownership to before nginx runs, this way nginx doesn't complain about not having permissions to this file.

**Comment by sravankasu** [comment]
Set [S6_READ_ONLY_ROOT](https://github.com/just-containers/s6-overlay/blob/master/README.md#read-only-root-filesystem) flag which helps get rid of `chmod` permission issues, but there seems to be no way around `s6-chown: fatal: unable to chown`, I looked at the [datadog-agent image](https://github.com/DataDog/datadog-agent/blob/master/Dockerfiles/agent/Dockerfile) and they have the same issue too.

**Comment by bryanlatten** [comment]
@sravankasu yes - I definitely thought that was a flag somewhere. What does the updated run look like? 
This might be worth an issue or even contribution upstream!

**Comment by sravankasu** [comment]
@bryanlatten It looks like this:

```
[init] executing /run.d/99-nginx.sh
[run] enabling web server
[run] starting process manager
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/10-nginx-config.sh: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/cont-init.d/11-nginx-validation.sh: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/cont-finish.d/00-nginx.sh: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/nginx/run: Operation not permitted
s6-chown: fatal: unable to chown /var/run/s6/etc/services.d/nginx/finish: Operation not permitted
exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-nginx-config.sh: executing... 
[nginx] adding app name (docker-ubuntu) to log format
[nginx] setting worker processes 1
[nginx] enabling minimal logging
[nginx] setting large_client_header_buffers to 8 16k
[nginx] setting client_header_buffer_size to 8k
[nginx] setting client_body_buffer_size to 128k
[cont-init.d] 10-nginx-config.sh: exited 0.
[cont-init.d] 11-nginx-validation.sh: executing... 
[cont-init.d] 11-nginx-validation.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:11
[services.d] done.
```

Yes, I can try opening an issue and see if they are open to addressing this.

**Comment by sravankasu** [comment]
Also looks like travis-ci doesn't like `COPY --chown`

**Comment by sravankasu** [comment]
Created an issue here: https://github.com/just-containers/s6-overlay/issues/272 for discussion.

File: container/root/scripts/set_permissions.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+touch /var/log/nginx/error.log
+
+# Assign ownership of required scripts/directories to the non root user
+rm /var/run && mkdir -p /var/run/s6 && chown -R ${NOT_ROOT_USER} /etc/services.d/ /etc/services-available/ /bin/s6-* /etc/nginx/ /tmp/ /var/
+
+chmod 755 -R /etc/s6/ /bin/s6-* /scripts/s6-setuidgid /etc/nginx/ /var/
+
+# Make s6 give non root user ownership to user provided files instead of root
+# List of files/folders given ownership available here: https://github.com/just-containers/s6-overlay/blob/master/builder/overlay-rootfs/etc/s6/init/init-stage2-fixattrs.txt
+sed -i "s/root/${NOT_ROOT_USER}/" /etc/s6/init/init-stage2-fixattrs.txt
```
Comment (sravankasu) [inline container/root/scripts/set_permissions.sh]: @bryanlatten Not sure why I didn't think of this earlier but this fixes all our issues with `s6-chown` and `s6-chmod` during `init-stage2`.

**Comment by sravankasu** [comment]
@bryanlatten PR is ready for review.

File: Dockerfile-alpine [hunk 4]
```diff
@@ -33,13 +33,15 @@ COPY ./container/root /
 # - Remove older WOFF mime-type
 # - Add again with newer mime-type
 # - Also add mime-type for WOFF2
+# Set permissions to allow image to be run under a non root user
```
Comment (bryanlatten) [inline Dockerfile-alpine]: can you combine this comment with line 29?

File: README.md [hunk 5]
```diff
@@ -18,6 +18,7 @@ See parent(s) [docker-base](https://github.com/behance/docker-base) for addition
 ### Expectations
 
 - Applications must copy their html/app into the `/var/www/html` folder
+- Any new script/file that needs to be added must be given proper permissions/ownership to the non root user through `container/root/scripts/set_permissions.sh`. This is to ensure that the image can be run under a non root user.
```
Comment (bryanlatten) [inline README.md]: 👍

**Comment by bryanlatten** [comment]
@sravankasu one small comment, then looks good!

**Comment by bryanlatten** [comment]
@sravankasu now just need passing tests...

**Comment by sravankasu** [comment]
@bryanlatten Done, per this ticket: https://github.com/moby/moby/issues/35018 using env vars in `COPY` is supported only from docker version: 19.03.

**Comment by bryanlatten** [comment]
and now we find out if docker hub supports it

**Pull request merged**